File: Python/calculator.py
from typing import Callable

from maybe import T
from parser import Parser, between
from some_parsers import identifier, symbol, double, usign

from math import sin, cos, asin, acos, sinh, cosh, asinh, acosh, tan, log, log10, exp, sqrt, pi, e

class Calculator(object):
    def __init__(self):
        pass

    def expr_in_brackets(self) -> Parser[float]:
        return between(symbol('('), symbol(')'), self.expr)
    
    def factor0(self) -> Parser[float]:
        return self.expr_in_brackets() | \
               self.funcs * self.expr_in_brackets | \
               self.consts | \
               double
    
    def factor(self) -> Parser[float]:
        return self.factor0().chainr1(self.fpow)
    
    def term(self) -> Parser[float]:
        return self.factor().chainl1(self.mul | self.div, False)
    
    def expr(self) -> Parser[float]:
        return usign.and_then(lambda sgn: self.term().chainl1(self.add | self.sub, sgn == "-"))
    
    def op2(c: str, f: Callable[[float, float], float]) -> Parser[Callable[[float, float], float]]:
        return symbol(c) >> (lambda: Parser.pure(f))
    
    add = op2('+', lambda x, y: x + y)
    sub = op2('-', lambda x, y: x - y)
    mul = op2('*', lambda x, y: x * y)
    div = op2('/', lambda x, y: x / y)
    fpow = op2('^', pow)

    def guard(b: bool, value: T) -> Parser[T]:
        return Parser.pure(value) if b else Parser.empty()

    funcs = identifier.and_then(lambda n:
        Calculator.guard(n == "sin",   sin)   |
        Calculator.guard(n == "cos",   cos)   |
        Calculator.guard(n == "asin",  asin)  |
        Calculator.guard(n == "acos",  acos)  |
        Calculator.guard(n == "sinh",  sinh)  |
        Calculator.guard(n == "cosh",  cosh)  |
        Calculator.guard(n == "asinh", asinh) |
        Calculator.guard(n == "acosh", acosh) |
        Calculator.guard(n == "tan",   tan)   |
        Calculator.guard(n == "log",   log)   |
        Calculator.guard(n == "log10", log10) |
        Calculator.guard(n == "exp",   exp)   |
        Calculator.guard(n == "sqrt",  sqrt)  |
        Calculator.guard(n == "sqr",   lambda x: x * x))

    consts = identifier.and_then(lambda n:
        Calculator.guard(n == "E",        e)          |
        Calculator.guard(n == "PI",       pi)         |
        Calculator.guard(n == "LOG2E",    1.44269504088896340736)  | # log2(e)
        Calculator.guard(n == "LOG10E",   0.434294481903251827651) | # log10(e)
        Calculator.guard(n == "LN2",      0.693147180559945309417) | # ln(2)
        Calculator.guard(n == "LN10",     2.30258509299404568402)  | # ln(10)
        Calculator.guard(n == "PI_2",     1.57079632679489661923)  | # pi/2
        Calculator.guard(n == "PI_4",     0.785398163397448309616) | # pi/4
        Calculator.guard(n == "1_PI",     0.318309886183790671538) | # 1/pi
        Calculator.guard(n == "2_PI",     0.636619772367581343076) | # 2/pi
        Calculator.guard(n == "2_SQRTPI", 1.12837916709551257390)  | # 2/sqrt(pi)
        Calculator.guard(n == "SQRT2",    1.41421356237309504880)  | # sqrt(2)
        Calculator.guard(n == "SQRT1_2",  0.707106781186547524401))  # 1/sqrt(2)

def calculate(inp):
    return Calculator().expr().parse(inp)

File: Ruby/lib/Calculator.rb
require_relative 'SomeParsers.rb'

def op2(c, &f)
    symbol(c).skip { Parser.pure(f.call) }
end

def add = op2('+') { lambda { |x, y| x + y } }
def sub = op2('-') { lambda { |x, y| x - y } }
def mul = op2('*') { lambda { |x, y| x * y } }
def div = op2('/') { lambda { |x, y| x / y } }
def pow = op2('^') { lambda { |x, y| Math.exp(y * Math.log(x)) } }

def guard(b, &x)
    b ? Parser.pure(x.call) : Parser.empty
end

def funcs = identifier.and_then { |n|
    guard(n == "sin")   { lambda { |x| Math.sin(x) } }   |
    guard(n == "cos")   { lambda { |x| Math.cos(x) } }   |
    guard(n == "asin")  { lambda { |x| Math.asin(x) } }  |
    guard(n == "acos")  { lambda { |x| Math.acos(x) } }  |
    guard(n == "sinh")  { lambda { |x| Math.sinh(x) } }  |
    guard(n == "cosh")  { lambda { |x| Math.cosh(x) } }  |
    guard(n == "asinh") { lambda { |x| Math.asinh(x) } } |
    guard(n == "acosh") { lambda { |x| Math.acosh(x) } } |
    guard(n == "tan")   { lambda { |x| Math.tan(x) } }   |
    guard(n == "atan")  { lambda { |x| Math.atan(x) } }  |
    guard(n == "log")   { lambda { |x| Math.log(x) } }   |
    guard(n == "log10") { lambda { |x| Math.log10(x) } } |
    guard(n == "exp")   { lambda { |x| Math.exp(x) } }   |
    guard(n == "sqrt")  { lambda { |x| Math.sqrt(x) } }  |
    guard(n == "sqr")   { lambda { |x| x * x } }
}

def consts = identifier.and_then { |n|
    guard(n == "E")        { 2.71828182845904523536  } |
    guard(n == "PI")       { 3.14159265358979323846  } |
    guard(n == "LOG2E")    { 1.44269504088896340736  } | # log2(e)
    guard(n == "LOG10E")   { 0.434294481903251827651 } | # log10(e)
    guard(n == "LN2")      { 0.693147180559945309417 } | # ln(2)
    guard(n == "LN10")     { 2.30258509299404568402  } | # ln(10)
    guard(n == "PI_2")     { 1.57079632679489661923  } | # pi/2
    guard(n == "PI_4")     { 0.785398163397448309616 } | # pi/4
    guard(n == "1_PI")     { 0.318309886183790671538 } | # 1/pi
    guard(n == "2_PI")     { 0.636619772367581343076 } | # 2/pi
    guard(n == "2_SQRTPI") { 1.12837916709551257390  } | # 2/sqrt(pi)
    guard(n == "SQRT2")    { 1.41421356237309504880  } | # sqrt(2)
    guard(n == "SQRT1_2")  { 0.707106781186547524401 }   # 1/sqrt(2)
}

def expr = usign.and_then { |sgn| chainl1(term, add | sub, sgn == "-") }
def term = chainl1(factor, mul | div, false)
def factor = factor0.chainr1(pow)

def factor0 = expr_in_brackets |
    funcs.apply { expr_in_brackets } |
    consts |
    double

def expr_in_brackets = between(symbol('('), symbol(')')) { expr }

def calculate(inp)
    expr.parse(inp)
end
